src/watertap_contrib/reflo/costing/units/forward_osmosis_zo.py

```diff
@@ -0,0 +1,153 @@
+# and should be applied to a Trevi FO flowsheet.
+# Mass and energy flows should be calculated from a Trevi FO flowsheet.
+
+
+def build_forward_osmosis_cost_param_block(blk):
+
+    costing = blk.parent_block()
+
+    blk.base_unit_cost = pyo.Var(
+        initialize=26784,
+        units=costing.base_currency / (pyo.units.m**3 / pyo.units.day),
+        bounds=(0, None),
+        doc="Base price of Trevi FO system ($/m3/day), including membranes, heat exchangers, \
+            construction, draw solution, coalescers, structural, polishing, pipes, plumbing, \
+            pre-filtration, control, pumps, instrumentation, valves, CIP and tanks",
+    )
+
+    blk.unit_cost_index = pyo.Var(
+        initialize=-0.428,
+        units=pyo.units.dimensionless,
+        bounds=(None, 0),
+        doc="Scaling factor of Trevi FO system capital cost",
+    )
+
+    blk.base_labor_cost = pyo.Var(
+        initialize=0.4757,
+        units=costing.base_currency / pyo.units.m**3,
+        bounds=(0, None),
+        doc="Base price of the labor cost ($/m3)",
+    )
+
+    blk.labor_cost_index = pyo.Var(
+        initialize=-0.178,
+        units=pyo.units.dimensionless,
+        bounds=(None, 0),
+        doc="Scaling factor of Trevi FO system capital cost",
+    )
+
+    blk.specific_energy_consumption_electric = pyo.Var(
+        initialize=1,
+        units=pyo.units.kWh / pyo.units.m**3,
+        bounds=(0, None),
+        doc="Specific electric energy consumption (kWh/m3)",
+    )
+
+    blk.cost_chemicals = pyo.Var(
+        initialize=0.07,
+        units=costing.base_currency / pyo.units.m**3,
+        bounds=(0, None),
+        doc="Cost of chemicals per m3 product",
+    )
+
+    blk.cost_durable_goods = pyo.Var(
+        initialize=0.05,
+        units=costing.base_currency / pyo.units.m**3,
+        bounds=(0, None),
+        doc="Cost of durable goods per m3 product",
+    )
+
+    blk.cost_disposal = pyo.Var(
+        initialize=0.02,
+        units=costing.base_currency / pyo.units.m**3,
+        bounds=(0, None),
+        doc="Cost of disposal per m3 brine",
+    )
+
+    blk.cost_fraction_insurance = pyo.Var(
+        initialize=0.005,
+        units=pyo.units.dimensionless,
+        bounds=(0, None),
+        doc="Fraction of capital cost for insurance",
+    )
+
+
+@register_costing_parameter_block(
+    build_rule=build_forward_osmosis_cost_param_block,
+    parameter_block_name="forward_osmosis",
+)
+def cost_forward_osmosis(blk):
+
+    fo_params = blk.costing_package.forward_osmosis
+    make_capital_cost_var(blk)
+    make_fixed_operating_cost_var(blk)
+
+    fo = blk.unit_model
+    test_fs = fo.flowsheet()
+
+    brine = fo.brine_props[0]
+
+    blk.costing_package.add_cost_factor(blk, None)
+    blk.annual_dist_production = pyo.units.convert(
+        test_fs.system_capacity, to_units=pyo.units.m**3 / pyo.units.year
+    )
+    blk.capital_cost_constraint = pyo.Constraint(
+        expr=blk.capital_cost
```

Comment:
If all costing params are indeed for ```pyunits.USD_2021```, you need to convert all costing variables to ```base_currency``` (and ideally ```base_period``` though this is always year to this point, it is not clear to me if defining the ```base_period``` to be something besides year like the duration of a multiperiod flowsheet would be useful for costing multiperiod flowsheets on an annual basis).

Reply:
For steady-state operation, I think converting to `base_period == pyunits.year` or other units should not impact anything. If it's variable operation, I guess it might overestimate the cost if using year within an hourly/daily multiperiod fs, so I'd use `base_currency / flow_rate` for those cases.
For this model, the variables are in `pyunits.USD_2021` now and I assume the solver will just convert those values to whatever the `base_currency` is?

Reply:
Yes correct. We always want to make sure we are converting our costing output to the ```base_currency``` set at the flowsheet level.

Reply:
Sorry @Zhuoran29, I think I skimmed/misunderstood your question. 

The solver doesn't do any unit conversions. The ```FlowsheetBlockCosting``` IDAES class the WaterTAP costing pkg inherits (which we then inherit) will convert all aggregated costs to ```base_currency``` and ```base_period``` units. So while it is true that we don't need to make the conversion here for the units to be correct on the flowsheet level, if we report costing at the unit process level (which seems likely) we will have to then convert all those values to the ```base_currency```. This is something that I typically forget to do (and check for when reviewing), but it has burned me in the past.

Reply:
Just remembered to do this: now `base_period` and `base_currency` are applied to the costing variables. Thanks for the clarification!